2.2-taptweak.ipynb

```diff
@@ -210,37 +217,101 @@
+    "# Alice modifies the contract and produces an alternative tweak\n",
+    "alternative_contract = \"Alice pays 0.1 BTC to Bob\"\n",
+    "t2 = sha256(alternative_contract.encode('utf-8'))\n",
+    "print(\"Tweak from original contract: {}\".format(t.hex()))\n",
+    "print(\"Tweak from modified contract: {}\\n\".format(t2.hex()))\n",
+    "\n",
+    "# Alice modifies her original secret and public key\n",
+    "# x` = x - t' + t\n",
+    "x_int = x_key.as_int()\n",
+    "t_int = int.from_bytes(t, \"big\") \n",
+    "t2_int = int.from_bytes(t2, \"big\") \n",
+    "x2_int = (x_int - t2_int + t_int) % SECP256K1_ORDER\n",
+    "x2_key = ECKey().set(x2_int, True)\n",
+    "P2_key = x2_key.get_pubkey()\n",
+    "\n",
+    "# The resulting tweaked public Q key remains the same\n",
+    "Q2_key = P2_key.tweak_add(t2)\n",
+    "print(\"Tweaked public key from original tweak t: {}\".format(Q_key.get_bytes().hex()))\n",
+    "print(\"Tweaked public key from modified tweak t2: {}\\n\".format(Q2_key.get_bytes().hex()))\n",
+    "\n",
+    "# So Alice can still produce a valid signature for Q\n",
+    "msg2 = sha256(b'I agree to the committed contract')\n",
+    "sig2 = q_key.sign_schnorr(msg2)\n",
+    "\n",
+    "# Given P2, Q and the modified contract, \n",
+    "# Alice can demonstrate the following to invalidate the original contract:\n",
+    "\n",
+    "# 1) The modified contract 'Alice pays 0.1 BTC to Bob' is appears to be committed to public key Q\n",
+    "print(\"'Alice pays 0.1 BTC to Bob' appears to be committed to Q: {}\"\\\n",
+    "      .format(P2_key.tweak_add(sha256(\"Alice pays 0.1 BTC to Bob\".encode('utf-8'))) == Q_key))\n",
+    "\n",
+    "# 2) Alice is still the owner of the public key Q = P2 + t2*G\n",
+    "print(\"Alice has produced a valid signature for Q: {}\".format(Q_key.verify_schnorr(sig2, msg)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example 2.2.4 - Pay-to-contract: Tweaking the pubkey with `H(P|msg)`\n",
+    "\n",
+    "In this example, we demonstrate a _secure_ commitment scheme called pay-to-contract. The private key is tweaked with the scalar `H(P|c)`. Since `P` appears both inside and outside the hash, it isn't possible to solve for a different `c` by modifying `x'`.\n",
     "\n",
-    "# Compute the tweak from H(P|msg)\n",
-    "commitment = b'commitment'\n",
-    "ss = sha256(pubkey.get_bytes())\n",
-    "ss += sha256(commitment)\n",
+    "* Alice can now no longer invalidate her previous contract commitment with Bob."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Alice generates a key pair\n",
+    "x_key, P_key = generate_key_pair()\n",
+    "print(\"Private key: {}\\nPublic key: {}\\n\".format(x_key.secret, P_key.get_bytes().hex()))\n",
+    "\n",
+    "# Alice compute the tweak from H(P|msg)\n",
+    "contract_bytes = \"Alice pays 10 BTC to Bob\".encode('utf-8')\n",
```

Comment:
This is inconsistent with the previous sections, should standardized on one approach